table.go

```diff
@@ -128,24 +240,41 @@ func (rt *RoutingTable) ResetCplRefreshedAtForID(id ID, newTime time.Time) {
 	rt.cplRefreshedAt[uint(cpl)] = newTime
 }
 
-// Update adds or moves the given peer to the front of its respective bucket
-func (rt *RoutingTable) Update(p peer.ID) (evicted peer.ID, err error) {
-	peerID := ConvertPeerID(p)
-	cpl := CommonPrefixLen(peerID, rt.local)
-
+// HandlePeerDisconnect should be called when the caller detects a disconnection with the peer.
+// This enables the Routing Table to mark the peer as missing.
+func (rt *RoutingTable) HandlePeerDisconnect(p peer.ID) {
 	rt.tabLock.Lock()
 	defer rt.tabLock.Unlock()
-	bucketID := cpl
-	if bucketID >= len(rt.Buckets) {
-		bucketID = len(rt.Buckets) - 1
+
+	// mark the peer as missing
+	bucketId := rt.bucketIdForPeer(p)
+	b := rt.Buckets[bucketId]
+	if peer, has := b.getPeer(p); has {
+		peer.State = PeerStateMissing
+		b.replace(peer)
 	}
+}
+
+// HandlePeerAlive should be called when the caller detects that a peer is alive.
+// This could be a successful incoming/outgoing connection with the peer or even a successful message delivery to/from the peer.
+// This enables the RT to update it's internal state to mark the peer as active.
+func (rt *RoutingTable) HandlePeerAlive(p peer.ID) (evicted peer.ID, err error) {
+	rt.tabLock.Lock()
+	defer rt.tabLock.Unlock()
 
+	bucketID := rt.bucketIdForPeer(p)
 	bucket := rt.Buckets[bucketID]
-	if bucket.Has(p) {
+	if peer, has := bucket.getPeer(p); has {
+		// mark the peer as active if it was missing
+		if peer.State == PeerStateMissing {
+			peer.State = PeerStateActive
+			bucket.replace(peer)
+		}
+
 		// If the peer is already in the table, move it to the front.
 		// This signifies that it it "more active" and the less active nodes
 		// Will as a result tend towards the back of the list
-		bucket.MoveToFront(p)
+		bucket.moveToFront(p)
```

Comment:
Having to update the peer's location twice is a bit unfortunate.

Reply:
@Stebalien You know, we don't actually need this code. Nowhere in the RT do we make decisions based on the location/recently seen metric of the peer , we only sort based on XOR distance.

I am removing this code for now. Let me know if you have any concerns.

Reply:
I've made the change.

Reply:
In the future, we'll likely want to sort by latency. However, for that, we'll probably use a heap.

IIRC, we _used_ to evict the oldest node when we found a new node. We've removed that and now _prefer_ the older nodes because that's the correct thing to do.